adapter/internal/sourcewatcher/sourcewatcher.go

```diff
@@ -0,0 +1,365 @@
+/*
+ *  Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package sourcewatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/wso2/product-microgateway/adapter/config"
+	"github.com/wso2/product-microgateway/adapter/internal/api"
+	xds "github.com/wso2/product-microgateway/adapter/internal/discovery/xds"
+	"github.com/wso2/product-microgateway/adapter/internal/loggers"
+	"github.com/wso2/product-microgateway/adapter/internal/oasparser/model"
+	"github.com/wso2/product-microgateway/adapter/pkg/auth"
+	"github.com/wso2/product-microgateway/adapter/pkg/logging"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+)
+
+const (
+	apisArtifactDir string = "apis"
+	zipExt          string = ".zip"
+)
+
+var artifactsMap map[string]model.ProjectAPI
+
+// Start fetches the API artifacts at the startup and polls for changes from the remote repository
+func Start() {
+	conf, err := config.ReadConfigs()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Error reading configs : %s", err.Error()),
+			Severity: logging.BLOCKER,
+			ErrorCode: 2500,
+		})
+	}
+
+	retryInterval := conf.Adapter.SourceControl.RetryInterval
+
+	loggers.LoggerSourceWatcher.Info("Starting source watcher")
+	// Fetch the API artifacts at the startup
+	repository, err := fetchArtifacts()
+
+	// Retry fetching the API artifacts if the first attempt fails
+	for {
+		if err == nil {
+			break
+		} else {
+			time.Sleep(time.Duration(retryInterval) * time.Second)
+			loggers.LoggerSourceWatcher.Info("Retrying fetching artifacts from the remote repository")
+			repository, err = fetchArtifacts()
+		}
+	}
+
+	artifactsMap, err = api.ProcessMountedAPIProjects()
+	if err != nil {
+		loggers.LoggerSourceWatcher.ErrorC(logging.ErrorDetails{
+			Message: fmt.Sprintf("Processing API artifacts failed : %s", err.Error()),
+			Severity: logging.CRITICAL,
+			ErrorCode: 2501,
+		})
+	}
+
+	loggers.LoggerSourceWatcher.Info("Polling for changes")
+	go pollChanges(repository)
+}
+
+// fetchArtifacts clones the API artifacts from the remote repository into the artifacts directory in the adapter
+func fetchArtifacts() (repository *git.Repository,err error) {
+	conf, err := config.ReadConfigs()
+	if err != nil {
```

Comment:
We can remove catching this error as it is already handled in the config package. (Therefore, the method should not return an error at all, it is something needs to be fixed in the existing code actually). You can do the same change for other locations as well.

Reply:
Fixed with 5eadc80